### PR TITLE
src/testing/benchmark.go: add implementation of Benchmark

### LINKED ISSUE #1808: Implement benchmark support in "tinygo test"
As https://github.com/tinygo-org/tinygo/blob/release/src/testing/benchmark.go notes, 
```
// TODO: Implement benchmarks. This struct allows test files containing
// benchmarks to compile and run, but will not run the benchmarks themselves.
```

### PR DESCRIPTION
**Pull request by dkegel-fastly**
Partially fixes #1808

### CODE REVIEW

**Comment by deadprogram** [comment]
Hi @dkegel-fastly thank you for this contribution.

So, how exactly should I go about testing it? In which environments? Thanks.

**Comment by dkegel-fastly** [comment]
I think it's portable, so any environment should be fine for testing.

A publicly available sample use might be:
```
curl "https://golang.org/test/fibo.go?m=text" > fibo.go
tinygo build -o fibo fibo.go
./fibo -bench
```
but alas that hits an internal error
```
LLVM ERROR: Cannot select: intrinsic %llvm.coro.size
```
at the moment.

**Comment by aykevl** [comment]
Not sure what the exact cause is of that error, but https://github.com/tinygo-org/tinygo/pull/1445 should fix it.

**Comment by deadprogram** [comment]
Hello @dkegel-fastly now that #1445 has been merged, perhaps you can take another look at this PR?

Thanks!

**Comment by fgsch** [comment]
I tried this (after fixing the Makefile conflict) and I'm seeing:

```
$ ./fibo -bench
panic: runtime error: nil pointer dereference
[1]    14689 abort      ./fibo -bench
```

With the default gc and scheduler under macOS. 
If I try the coroutines scheduler I'm back at `LLVM ERROR: Cannot select: intrinsic %llvm.coro.size`.

**Comment by aykevl** [comment]
Can you try running this binary under a debugger (gdb/lldb)? That should show on which line it is dereferencing a nil pointer. You can set a breakpoint on `runtime._abort`.

**Comment by fgsch** [comment]
@aykevl sure, I will try shortly and add the info here.

**Comment by fgsch** [comment]
Backtrace:

```
#1  0x00007ffff7df3859 in __GI_abort () at abort.c:79
#2  0x00000000004014f6 in runtime.runtimePanic (msg=...) at /home/fgsch/tinygo/src/runtime/panic.go:20
#3  0x00000000004017f4 in runtime.nilPanic () at /home/fgsch/tinygo/src/runtime/panic.go:32
#4  0x0000000000401c10 in (*flag.FlagSet).Var (f=0x0, value=..., name=..., usage=...) at /opt/go-1.16.4/src/flag/flag.go:862
#5  0x0000000000401a73 in runtime.initAll () at /home/fgsch/tinygo/src/runtime/runtime.go:33
#6  0x0000000000401a42 in runtime.run$1 () at /home/fgsch/tinygo/src/runtime/scheduler_any.go:19
#7  0x0000000000401a2c in <goroutine wrapper> () at /home/fgsch/tinygo/src/runtime/scheduler_any.go:18
#8  0x0000000000401c31 in tinygo_startTask () at /home/fgsch/tinygo/src/internal/task/task_stack_amd64.S:29
```

**Comment by aykevl** [comment]
Huh, interesting. This might be a bug in the interp package. Can you rebase and force-push the PR so I can debug this locally?

**Comment by fgsch** [comment]
It's not my PR so we'll need to ask @dkegel-fastly. You can ignore the Makefile change in this PR and instead do:

```
diff --git a/Makefile b/Makefile
index e2981789..7b15d088 100644
--- a/Makefile
+++ b/Makefile
@@ -207,6 +207,8 @@ TEST_PACKAGES = \
 .PHONY: tinygo-test
 tinygo-test:
        $(TINYGO) test $(TEST_PACKAGES)
+       # until "test testing" passes
+       cd tests/tinygotest && $(TINYGO) test benchmark_test.go

 .PHONY: smoketest
 smoketest:
```

**Comment by dkegel-fastly** [comment]
Rebased.  After a clean build of go, building fibo.go with tinygo still yields
```
panic: runtime error: nil pointer dereference
Abort trap: 6
```

**Comment by fgsch** [comment]
This should be good to go in now.

File: src/testing/benchmark.go [hunk 1]
```diff
@@ -6,29 +6,209 @@
...
+        if r.N <= 0 {
+                return 0
+        }
+        return r.T.Nanoseconds() / int64(r.N)
+}
+
+// AllocsPerOp returns the "allocs/op" metric,
+// which is calculated as r.MemAllocs / r.N.
+func (r BenchmarkResult) AllocsPerOp() int64 {
+        return 0	// Dummy version to allow running fibo.go
```
Comment (peterbourgon) [inline src/testing/benchmark.go]: gofmt

File: src/testing/benchmark.go [hunk 2]
```diff
@@ -6,29 +6,209 @@
...
 type InternalBenchmark struct {
 	Name string
 	F    func(b *B)
 }
 
+// BenchmarkResult contains the results of a benchmark run.
+type BenchmarkResult struct {
+        N                   int           // The number of iterations.
+        T                   time.Duration // The total time taken.
+}
```
Comment (peterbourgon) [inline src/testing/benchmark.go]: gofmt

File: src/testing/benchmark.go [hunk 3]
```diff
@@ -6,29 +6,209 @@
...
 }
 
+// runN runs a single benchmark for the specified number of iterations.
+func (b *B) runN(n int) {
+	b.N = n
+        b.ResetTimer()
+        b.StartTimer()
+        b.benchFunc(b)
+        b.StopTimer()
+}
```
Comment (peterbourgon) [inline src/testing/benchmark.go]: gofmt

Comment (dkegel-fastly) [reply src/testing/benchmark.go]: Done

**Comment by deadprogram** [comment]
Looks good to me now, thanks for the updates @dkegel-fastly 

@aykevl any more feedback before merge?

File: Makefile [hunk 4]
```diff
@@ -207,6 +207,8 @@ TEST_PACKAGES = \
 .PHONY: tinygo-test
 tinygo-test:
 	$(TINYGO) test $(TEST_PACKAGES)
+	# until "test testing" passes
+	cd tests/tinygotest && $(TINYGO) test benchmark_test.go
```
Comment (aykevl) [inline Makefile:211]: I wonder why this doesn't work. Ideally I'd like to see this fixed first but maybe it's better to merge this PR first so it doesn't stall any longer.

File: src/testing/benchmark.go [hunk 5]
```diff
@@ -6,29 +6,209 @@
...
+	if r.N <= 0 {
+		return 0
+	}
+	return r.T.Nanoseconds() / int64(r.N)
+}
+
+// AllocsPerOp returns the "allocs/op" metric,
+// which is calculated as r.MemAllocs / r.N.
+func (r BenchmarkResult) AllocsPerOp() int64 {
+	return 0 // Dummy version to allow running fibo.go
```
Comment (aykevl) [inline src/testing/benchmark.go]: What is this fibo.go file? It is also mentioned in `AllocedBytesPerOp`.

File: tests/tinygotest/benchmark_test.go [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+// TestBenchmark simply uses Benchmark twice and makes sure it does not crash.
+func TestBenchmark(t *testing.T) {
+	// FIXME: reduce runtime from the current 3 seconds.
+	rslow := testing.Benchmark(BenchmarkSlowNonASCII)
+	rfast := testing.Benchmark(BenchmarkFastNonASCII)
+	tslow := rslow.NsPerOp()
+	tfast := rfast.NsPerOp()
+
+	//fmt.Printf("TestBenchmark: tslow %d ns, %f ms; tfast %d ns, %f ms\n", tslow, float64(tslow) / 1000000.0, tfast, float64(tfast) / 1000000.0)
```
Comment (aykevl) [inline tests/tinygotest/benchmark_test.go]: This looks like leftover debugging code? I think it should be removed.

**Review by aykevl — COMMENTED** [review]
Overall this looks good to me, with a few small nitpicks inline.

File: src/testing/benchmark.go (hunk 5 above)
Comment (dkegel-fastly) [reply src/testing/benchmark.go]: Added rest of the url.

File: tests/tinygotest/benchmark_test.go (hunk 6 above)
Comment (dkegel-fastly) [reply tests/tinygotest/benchmark_test.go]: Deleted.

File: Makefile (hunk 4 above)
Comment (dkegel-fastly) [reply Makefile:211]: Filed https://github.com/tinygo-org/tinygo/issues/2027

Comment (deadprogram) [reply Makefile:211]: > maybe it's better to merge this PR first so it doesn't stall any longer.

Agreed.

**Comment by deadprogram** [comment]
Looking good now, @dkegel-fastly thanks for making all the requested changes and thanks @peterbourgon @fgsch @aykevl for reviewing/feedback.

Merging.

**Pull request merged**